source/util/move_to_front.h

```diff
@@ -0,0 +1,318 @@
+//
+// Terminology
+// id: SPIR-V id.
+// rank: 0-indexed value showing how recently the id was accessed.
+// node: uint16_t handle used internally to access nodes.
+// size: size of the subtree of a node (including the node).
+// height: distance from a node to the farthest leaf.
+// index: not used to avoid confusion with rank and handle.
+class MoveToFront {
+  struct Node;
+ public:
+  explicit MoveToFront(size_t reserve_capacity = 128) {
+    nodes_.reserve(reserve_capacity);
+
+    // Create NIL node.
+    nodes_.emplace_back(Node());
+  }
+
+  // Returns the 0-indexed rank of id in the move-to-front sequence and moves
+  // id to the front. Example:
+  // Before the call: 4 8 2 1 7
+  // RankFromId(8) returns 1
+  // After the call: 8 4 2 1 7
+  size_t RankFromId(uint32_t id);
+
+  // Returns id corresponding to a 0-indexed rank in the move-to-front sequence
+  // and moves id to the front. Example:
+  // Before the call: 4 8 2 1 7
+  // IdFromRank(1) returns 8
+  // After the call: 8 4 2 1 7
+  uint32_t IdFromRank(size_t rank);
+
+  // Permanently removes the id from the move-to-front sequence.
+  void DeprecateId(uint32_t id) {
+    auto it = id_to_node_.find(id);
+    assert(it != id_to_node_.end());
+    RemoveNode(it->second);
+    // The iterator should still be valid, even if RemoveNode has modified
+    // id_to_node_. But just in case erase by id, not by iterator.
+    id_to_node_.erase(id);
+
+#ifndef NDEBUG
+    deprecated_ids_.insert(id);
+#endif
+  }
+
+  // Returns the number of elements in the move-to-front sequence.
+  size_t GetSize() const {
+    return SizeOf(root_);
+  }
+
+  // The methods below are for testing only.
+
+  // Inserts the value in the internal tree data structure. For testing only.
+  void TestInsert(uint32_t val) {
+    InsertNode(CreateNode(val, val));
+  }
+
+  // Removes the value from the internal tree data structure. For testing only.
+  void TestRemove(uint32_t val) {
+    const auto it = id_to_node_.find(val);
+    assert (it != id_to_node_.end());
+    RemoveNode(it->second);
+  }
+
+  // Prints the internal tree data structure to |out|. For testing only.
+  void PrintTree(std::ostream& out, bool print_timestamp = false) const {
+    if (root_)
+      PrintTreeInternal(out, root_, 1, print_timestamp);
+  }
+
+ private:
+  // Internal tree data structure node.
+  struct Node {
+    // Fields are in order from biggest bit width to smallest.
+    // SPIR-V id.
+    uint32_t id = 0;
+    // Timestamp from a logical clock which updates every time the element is
+    // accessed.
+    uint32_t timestamp = 0;
+    // The size of the node's subtee, including the node.
+    // SizeOf(LeftOf(node)) + SizeOf(RightOf(node)) + 1.
+    uint16_t size = 0;
+    // Handles to connected nodes.
+    uint16_t left = 0;
+    uint16_t right = 0;
+    uint16_t parent = 0;
+    // Distance to the farthest leaf.
+    // Leaves have height 0, real nodes at least 1.
+    uint8_t height = 0;
```

Comment:
We should run into hardware problems long before we reach tree height 255. But you have a point, I'm not saving anything at all, because after that field would be a gap, and after the gap the next struct.